saleor/graphql/product/types/products.py

```diff
@@ -612,6 +591,39 @@ def calculate_global_sold_units(variant_channel_listings):
 
         return variant_channel_listings.then(calculate_global_sold_units)
 
+    @staticmethod
+    def __resolve_references(roots: List["ProductVariant"], info, **_kwargs):
+        requestor = get_user_or_app_from_context(info.context)
+        requestor_has_access_to_all = has_one_of_permissions(
+            requestor, ALL_PRODUCTS_PERMISSIONS
+        )
+        channels = defaultdict(set)
+        roots_ids = []
+        for root in roots:
+            _, root_id = from_global_id_or_error(
+                root.id, ProductVariant, raise_error=True
+            )
+            if root_id:
+                roots_ids.append(f"{root.channel}_{root_id}")
+                channels[root.channel].add(root_id)
+
+        variants = {}
+        for channel, ids in channels.items():
+            visible_products = models.Product.objects.visible_to_user(
+                requestor, channel
+            ).values_list("pk", flat=True)
```

Comment:
Could we use `resolve_product_variants` here somehow? If possible we shouldn't duplicate the resolver logic here, especially the logic responsible for limiting visibility of products/variants to users based on their permission/ staff status etc.